lib/internal/test_runner/test.js

```diff
@@ -605,9 +605,23 @@ class Test extends AsyncResource {
           throw new ERR_INVALID_ARG_TYPE('options.concurrency', ['boolean', 'number'], concurrency);
     }
 
+    // Take timeout from cli
+    const cliTimeout = this.config.timeout;
+    if (cliTimeout != null && cliTimeout !== Infinity) {
+      if (!(this.name === '<root>' && this.parent === null) &&
```

Comment:
this `!(this.name === '<root>' && this.parent === null)` bit of logic is duplicated twice - I will refactor it. But just trying to make sure this makes sense first @cjihrig

Reply:
I don't think we should do this. Can you set `this.timeout = null;` or whatever needs to happen in the `FileTest` and `Suite` constructor after the `super()` calls? It would be similar to [this](https://github.com/nodejs/node/blob/b1402835a512f14fa9f8dd23d3e0cee8cfe888a2/lib/internal/test_runner/test.js#L1402).

Reply:
Yes, what you suggested works! Sorry, I just felt I am stupid 😓

Reply:
a couple of tests that use `run` start not to timeout - I will investigate why tomorrow 👍

Reply:
I think `this.timeout = null` set on `FileTest` makes `run({timeout: x})` lost effect

Reply:
That shouldn't be the case. A `FileTest` should never timeout. In the spawned child process, you should be able to log the `--test-timeout` flag to verify that it gets propagated properly.

Is it possible that the things that are no longer timing out are things that never should have timed out in the first place, but did because of the incorrect implementation?

Reply:
I think possibly it is related to the incorrect implementation. Here is why I think it is:

There were two tests timed out, I only had time to investigate one of them `test/parallel/test-runner-filetest-location.js`.

The reason why it is failing is because `timeout` used to set on the `FileTest` and after `100ms` the test execution would stop, but now because we reset `this.timeout = null` in `FileTest` constructor, the timeout option is not properly propagated to the per test level, the test is simply a `while true` loop which I guess cause the event loop blocked and promise never got a chance to resolve. Still looking but would be great to get some help 🙏

Reply:
for the reference, the failed test didn't use `--test-timeout` flag, the `timeout` from options prints `undefined`.

Reply:
I've just checked again, I think if user has a `while true` loop in their `test` function, it should be their responsibility to handle the event loop blocking not the test runner?  Not sure if what I've said makes sense though, please do let me know anything that confuse you.

I have even tried to manually pass in the `timeout` to child process but the `SafePromiseRace` still couldn't resolve which leads to hanging eventually --> test timeout.

Reply:
In another word - previously because test timeout is set on the `FileTest` so when running `stopTestPromise` and promise that the test that contains `while true` loop that runs in a child process, the main process is able to resolve the `stopTestPromise` in the Promise.race and cancel the test running in the child process.

Right now because the Promise.race cannot resolve due to the `while true` loop in the test blocks the event loop, the test runner just hangs.

I am not entirely sure how I can proceed with this, would be great to have some your input 🙏 @cjihrig

Reply:
`test/parallel/test-runner-filetest-location.js` is really only testing that locations are reported correctly, so I think it would be fine for that test to be refactored to not rely on timeouts or a `while` loop. It might be enough to run a file that includes no tests and exits with a non-zero exit code, but I'm not 100% sure without testing it myself.

I also agree that if a test enters a blocking infinite loop that we might not be able to do anything about that. For example, if you are running a single file without `--test` or with test isolation disabled, I don't think we'd be able to handle that case anyway.

Reply:
FWIW, I just verified that mocha does not time out on the following test either:

```js
it('infinite loop', () => {
  while (true);
});
```

Reply:
thanks for verifying!